src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalyzerAssemblyLoader.cs

```diff
@@ -346,83 +342,78 @@ public Assembly LoadFromPath(string originalAnalyzerPath)
-        /// </summary>
-        protected abstract string PreparePathToLoad(string assemblyFilePath);
-
-        /// <summary>
-        /// When overridden in a derived class, allows substituting a satellite assembly path after we've
-        /// identified the context to load a satellite assembly in, but before the satellite assembly is actually
-        /// loaded from disk. This is used to substitute out the original path with the shadow-copied version.
-        /// </summary>
-        protected abstract string PrepareSatelliteAssemblyToLoad(string assemblyFilePath, string cultureName);
-
-        /// <summary>
-        /// When <see cref="PreparePathToLoad(string)"/> is overridden this returns the most recent
-        /// real path calculated for the <paramref name="originalFullPath"/>
-        /// </summary>
-        internal string GetRealAnalyzerLoadPath(string originalFullPath)
+        internal (string OriginalAssemblyPath, string RealAssemblyPath)[] GetPathMapSnapshot()
         {
             CheckIfDisposed();
 
             lock (_guard)
             {
-                if (!_analyzerAssemblyInfoMap.TryGetValue(originalFullPath, out var tuple))
-                {
-                    throw new InvalidOperationException($"Invalid original path: {originalFullPath}");
-                }
-
-                return tuple is { } value ? value.RealAssemblyPath : originalFullPath;
+                return _realToOriginalPathMap.Select(x => (x.Value, x.Key)).ToArray();
             }
         }
 
-        internal (string OriginalAssemblyPath, string RealAssemblyPath)[] GetPathMapSnapshot()
+#if NET
+        /// <summary>
+        /// Return an <see cref="IAnalyzerAssemblyLoader"/> which does not lock assemblies on disk that is
+        /// most appropriate for the current platform.
+        /// </summary>
+        /// <param name="windowsShadowPath">A shadow copy path will be created on Windows and this value 
+        /// will be the base directory where shadow copy assemblies are stored. </param>
+        internal static IAnalyzerAssemblyLoaderInternal CreateNonLockingLoader(
+            string windowsShadowPath,
+            ImmutableArray<IAnalyzerPathResolver> pathResolvers = default,
+            ImmutableArray<IAnalyzerAssemblyResolver> assemblyResolvers = default,
+            System.Runtime.Loader.AssemblyLoadContext? compilerLoadContext = null)
         {
-            CheckIfDisposed();
+            pathResolvers = pathResolvers.NullToEmpty();
+            assemblyResolvers = assemblyResolvers.NullToEmpty();
 
-            lock (_guard)
+            if (!RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+            {
+                return new AnalyzerAssemblyLoader(
+                    pathResolvers,
+                    [.. assemblyResolvers, StreamResolver.Instance],
+                    compilerLoadContext);
+            }
```

Comment:
consider doc'ing why this is what we create for non-windows. 
consider doc'ing why the stream resolver instance is appropriate to pass in.